go/vt/vttablet/tabletmanager/semisyncmonitor/monitor_test.go

```diff
@@ -0,0 +1,714 @@
+	// Start a cancellable context and use that to wait.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	wg.Add(1)
+	var ctxErr error
+	var mu sync.Mutex
+	go func() {
+		defer wg.Done()
+		err := m.WaitUntilSemiSyncUnblocked(ctx)
+		mu.Lock()
+		ctxErr = err
+		mu.Unlock()
+	}()
+
+	// Start another go routine, also waiting for semi-sync being unblocked, but not using the cancellable context.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		err := m.WaitUntilSemiSyncUnblocked(context.Background())
+		require.NoError(t, err)
+	}()
+
+	// Wait until the writes have started.
+	require.Eventually(t, func() bool {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+		return m.isWriting
+	}, 1*time.Second, 100*time.Millisecond)
+
+	// Now we cancel the context. This should fail the first wait.
+	cancel()
+	// Since we cancel the context before the semi-sync has been unblocked, we expect a context timeout error.
+	require.Eventually(t, func() bool {
+		mu.Lock()
+		defer mu.Unlock()
+		return ctxErr != nil
+	}, 1*time.Second, 100*time.Millisecond)
+	mu.Lock()
+	require.EqualError(t, ctxErr, "context canceled")
+	mu.Unlock()
+
+	// Now we set the monitor to be unblocked.
+	db.AddQuery(semiSyncWaitSessionsRead, sqltypes.MakeTestResult(sqltypes.MakeTestFields("Variable_name|Value", "varchar|varchar"), "Rpl_semi_sync_source_wait_sessions|0"))
+	err = m.WaitUntilSemiSyncUnblocked(context.Background())
+	require.NoError(t, err)
+	// This should unblock the second wait.
+	wg.Wait()
+	// Eventually the writes should also stop.
+	require.Eventually(t, func() bool {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+		return !m.isWriting
+	}, 1*time.Second, 100*time.Millisecond)
+
+	// Also verify that if the monitor is closed, we don't wait.
+	m.Close()
+	err = m.WaitUntilSemiSyncUnblocked(context.Background())
+	require.NoError(t, err)
+	require.True(t, m.isClosed())
+}
+
+// TestSemiSyncMonitor tests the semi-sync monitor as a black box.
+// It only calls the exported methods to see they work as intended.
+func TestSemiSyncMonitor(t *testing.T) {
+	initialVal := waitBetweenWrites
+	waitBetweenWrites = 250 * time.Millisecond
+	defer func() {
+		waitBetweenWrites = initialVal
+	}()
+	db := fakesqldb.New(t)
+	defer db.Close()
+	params := db.ConnParams()
+	cp := *params
+	dbc := dbconfigs.NewTestDBConfigs(cp, cp, "")
+	config := &tabletenv.TabletConfig{
+		DB: dbc,
+		SemiSyncMonitor: tabletenv.SemiSyncMonitorConfig{
+			Interval: 1 * time.Second,
+		},
+	}
+	m := NewMonitor(config, exporter)
+	defer m.Close()
+
+	db.SetNeverFail(true)
+	// Initially everything is unblocked.
+	db.AddQuery(semiSyncWaitSessionsRead, sqltypes.MakeTestResult(sqltypes.MakeTestFields("Variable_name|Value", "varchar|varchar"), "Rpl_semi_sync_source_wait_sessions|0"))
+
+	// Open the monitor.
+	m.Open()
+
+	// Initially writes aren't blocked and the wait returns immediately.
+	require.False(t, m.AllWritesBlocked())
+	err := m.WaitUntilSemiSyncUnblocked(context.Background())
```

Comment:
We could use a context with a short timeout here to verify it returns immediately and to prevent the test from potentially hanging.

Reply:
Fixed